### PR TITLE
[#12048] data migration for whole course student chain

### PR DESCRIPTION
**Pull request by mingyuanc**
Part of https://github.com/TEAMMATES/teammates/issues/12048

Changes:
Cascade migrate course
Verify migrated course

### CODE REVIEW

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        // Runnable shutdownScript = () -> { cascadeDelete(newCourse)};
+        // Runtime.getRuntime().addShutdownHook(new Thread(shutdownScript));
+
+        migrateCourseEntity(newCourse);
+        // verifyCourseEntity(newCourse);
+        // markOldCourseAsMigrated(courseId)
+        // Runtime.getRuntime().removeShutDownHook(new Thread(shutdownScript));
+    }
+
+    private teammates.storage.sqlentity.Course createCourse(Course oldCourse) {
```
Comment (FergusMok) [inline src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:60]: With this new setup, will we be getting rid of the previous scripts? E.g `DataMigrationForCourseSql`

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    private void migrateCourseEntity(teammates.storage.sqlentity.Course newCourse) {
+        Map<String, teammates.storage.sqlentity.Section> sectionNameToSectionMap = migrateSectionChain(newCourse);
+        System.out.println(sectionNameToSectionMap); // To stop lint from complaining
+        // TODO: Add mirgrateFeedbackChain
+        // migrateFeedbackChain(sectionNameToSectionMap);
+    }
+
+    private Map<String, teammates.storage.sqlentity.Section> migrateSectionChain(
```
Comment (FergusMok) [inline src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:88]: Could we arrange the methods in order so it'll be clearer, whichever the order. E.g the main "chain" methods, then the helper methods, or vice versa.

File: src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                .collect(Collectors.groupingBy(CourseStudent::getSectionName));
+        Map<String, List<Student>> sectionToNewStuMap = newStudents.stream()
+                .collect(Collectors.groupingBy(Student::getSectionName));
+
+        List<Section> newSection = sqlEntity.getSections();
+
+        // Enforce that they are equal in size so the next check can check if all
+        // sections are present
+        if (newSection.size() != sectionToOldStuMap.size() || newSection.size() != sectionToNewStuMap.size()) {
+            return false;
```
Comment (FergusMok) [inline src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java:86]: One concern is that if verification fails, we should still be able to see which entity failed, and at which attribute / count. With this, will we still be able to see that? Similarly for the rest.

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            teammates.storage.sqlentity.Team newTeam = createTeam(newSection, teamName);
+            migrateStudent(newCourse, newTeam, stuList);
+            saveEntityDeferred(newTeam);
+        }
+    }
+
+    private void migrateStudent(teammates.storage.sqlentity.Course newCourse, teammates.storage.sqlentity.Team newTeam,
+            List<CourseStudent> studentsInTeam) {
+        for (CourseStudent oldStudent : studentsInTeam) {
+            teammates.storage.sqlentity.Student newStudent = new teammates.storage.sqlentity.Student(newCourse,
```
Comment (FergusMok) [inline src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:123]: Should we standardize with the `create..` methods too?

File: src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                .list();
+
+        HibernateUtil.beginTransaction();
+        CriteriaBuilder cb = HibernateUtil.getCriteriaBuilder();
+        CriteriaQuery<teammates.storage.sqlentity.Student> cr = cb
+                .createQuery(teammates.storage.sqlentity.Student.class);
+        Root<teammates.storage.sqlentity.Student> courseRoot = cr.from(teammates.storage.sqlentity.Student.class);
+        cr.select(courseRoot).where(cb.equal(courseRoot.get("courseId"), sqlEntity.getId()));
+        List<Student> newStudents = HibernateUtil.createQuery(cr).getResultList();
+        HibernateUtil.commitTransaction();
```
Comment (FergusMok) [inline src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java:74]: Could we abstract this out into another method too? E.g `getNewStudents`

File: src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        Map<String, List<CourseStudent>> sectionToOldStuMap = oldStudents.stream()
+                .collect(Collectors.groupingBy(CourseStudent::getSectionName));
+        Map<String, List<Student>> sectionToNewStuMap = newStudents.stream()
+                .collect(Collectors.groupingBy(Student::getSectionName));
+
+        List<Section> newSection = sqlEntity.getSections();
+
+        // Enforce that they are equal in size so the next check can check if all
+        // sections are present
+        if (newSection.size() != sectionToOldStuMap.size() || newSection.size() != sectionToNewStuMap.size()) {
```
Comment (FergusMok) [inline src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java:85]: Can do something similar to this, to make use of code as documentation, and do away with the comments. This way, we can see more clearly the parts where the count check was made too.

```suggestion
        boolean isSectionsCountEqual = (newSection.size() != sectionToOldStuMap.size() || newSection.size() != sectionToNewStuMap.size());
        if (!isSectionsCountEqual) {
            return false;
        }
```

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Override
+    protected void migrateEntity(Course oldCourse) throws Exception {
+        teammates.storage.sqlentity.Course newCourse = createCourse(oldCourse);
+        // TODO: add shutdown hook to save the entity
+        // Runnable shutdownScript = () -> { cascadeDelete(newCourse)};
+        // Runtime.getRuntime().addShutdownHook(new Thread(shutdownScript));
+
+        migrateCourseEntity(newCourse);
+        // verifyCourseEntity(newCourse);
```
Comment (FergusMok) [inline src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:55]: I'm assuming if verification fails, somehow the cascadeDelete will happen too?

**Review by FergusMok — COMMENTED** [review]
Good progress, but some important questions about the overall design that the others in the team can chime in about too

**Comment by NicolasCwy** [comment]
Agree with Fergus, could you abstract some methods out?

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java (hunk 1 above)
Comment (mingyuanc) [reply src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:60]: I believe so

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java (hunk 2 above)
Comment (mingyuanc) [reply src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:88]: Sure will do!

File: src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java (hunk 3 above)
Comment (mingyuanc) [reply src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java:86]: Good question, let me discuss with my teammates

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java (hunk 4 above)
Comment (mingyuanc) [reply src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:123]: Good catch will do!

File: src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java (hunk 5 above)
Comment (mingyuanc) [reply src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java:74]: Sure good catch!

File: src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java (hunk 6 above)
Comment (mingyuanc) [reply src/client/java/teammates/client/scripts/sql/VerifyCourseEntityAttributes.java:85]: Sure good catch!

File: src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java (hunk 7 above)
Comment (mingyuanc) [reply src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java:55]: Let me check with my teammates

**Comment by NicolasCwy** [comment]
To make this easier to review perhaps you can split each step into separate PRs so we can review each step one by one starting with the one with the least dependencies

**Comment by NicolasCwy** [comment]
Merged in #13036

**Pull request closed without merging**